# Ejemplos Nueva Arquitectura (Actualizado)

Flujo completo del Portfolio Tracker con la arquitectura multi-proveedor:

| Sección | Descripción |
|---|---|
| **1. Inicializar Portfolio** | Carga Excel de Órdenes + FIFO automático |
| **2. Posiciones Activas** | Valoración en tiempo real (FMP → YFinance) |
| **3. Enriquecimiento de datos** | 6 bloques independientes con benchmark integrado |
| **4. Calculadora Fiscal** | Retirada optimizada con mínimo impacto fiscal |

**Bloques de enriquecimiento (Sección 3):**

| # | Bloque | Proveedor principal | Fallback |
|---|--------|---------------------|----------|
| 3.1 | Comisiones | **Finect** (INITIAL_STATE JSON, ~800ms/1ª vez, luego caché) | FT → FMP |
| 3.2 | Sectores | **Finect** (stock-sector, caché) | FT → FMP |
| 3.3 | Regiones | **Finect** (regional-exposure, caché) | FT → FMP |
| 3.4 | Asset Allocation | **Finect** (asset-allocation, caché) | FT → YFinance |
| 3.5 | Portfolio (Holdings) | **Finect** (portfolio.holdings, caché) | FT → YFinance → FMP |
| 3.6 | Estrellas Morningstar | **Morningstar** (mstarpy, única fuente de rating) | — |

> **Ventaja clave de Finect**: Una sola petición HTTP por fondo (~800ms) obtiene *todos* los bloques 3.1–3.5.
> Las llamadas subsiguientes al mismo ISIN son instantáneas (caché en sesión).
> FT, YFinance y FMP actúan como fallback para fondos no indexados en Finect.


---
## 1. Inicializar Portfolio desde Excel

In [13]:
import sys, os, warnings
import pandas as pd
from IPython.display import display, HTML

warnings.filterwarnings('ignore')

# Resolver rutas según desde dónde se lance el notebook
if os.path.exists('../app'):
    sys.path.append(os.path.abspath('..'))
    path = '../data/Ordenes.xlsx'
else:
    sys.path.append(os.path.abspath('./backend'))
    path = './backend/data/Ordenes.xlsx'

from app.services.core_portfolio import Portfolio

portfolio = Portfolio(path)
isins_cartera = list(portfolio.positions.keys())

print(f"✅ Portfolio inicializado con éxito.")
print(f"   Lotes abiertos: {len(portfolio.get_open_lots())}")
print(f"   Fondos en cartera: {len(isins_cartera)}")
print(f"   ISINs: {isins_cartera}")

✅ Portfolio inicializado con éxito.
   Lotes abiertos: 61
   Fondos en cartera: 18
   ISINs: ['ES0140794001', 'ES0141116030', 'ES0146309002', 'IE000ZYRH0Q7', 'IE00B88WFS66', 'IE00BD0NCM55', 'IE00BH6XSF26', 'IE00BYX5MX67', 'IE00BYX5NX33', 'LU0273159177', 'LU0302296495', 'LU0329355670', 'LU0840158819', 'LU1598719752', 'LU1694789451', 'LU1988110927', 'LU2466448532', 'LU3038481936']


---
## 2. Posiciones Activas

NAV obtenido vía cadena rápida: **FMP → YFinance** (sin esperas de scraping).

In [4]:
print("Obteniendo valoraciones...")
df_posiciones = portfolio.get_positions()
display(df_posiciones)

print(f"\n  Fondos en cartera:  {len(df_posiciones)}")
print(f"  Capital Invertido:  {portfolio.get_total_invested():,.2f} €")

Obteniendo valoraciones...
Cache manager initialized with base path: c:\Users\jaguirrepeman\OneDrive - Deloitte (O365D)\Documents\DS\Finance\backend\app\data\cache
  ✓ Usando datos en caché para ES0140794001 (light)
Cache manager initialized with base path: c:\Users\jaguirrepeman\OneDrive - Deloitte (O365D)\Documents\DS\Finance\backend\app\data\cache
  ✓ Usando datos en caché para ES0141116030 (light)
Cache manager initialized with base path: c:\Users\jaguirrepeman\OneDrive - Deloitte (O365D)\Documents\DS\Finance\backend\app\data\cache
  ✓ Usando datos en caché para ES0146309002 (light)
Cache manager initialized with base path: c:\Users\jaguirrepeman\OneDrive - Deloitte (O365D)\Documents\DS\Finance\backend\app\data\cache
  ✓ Usando datos en caché para IE000ZYRH0Q7 (light)
Cache manager initialized with base path: c:\Users\jaguirrepeman\OneDrive - Deloitte (O365D)\Documents\DS\Finance\backend\app\data\cache
  ✓ Usando datos en caché para IE00B88WFS66 (light)
Cache manager initialized wi

,ISIN,Fondo,Participaciones,Valor_Euros,Fecha_Valoracion
0,IE00BYX5MX67,Fidelity Ucits II Icav - Fidelity S&P 500 Inde...,1989.463,25411.41,2025-06-11
1,IE00BD0NCM55,iShares Dev Wld Idx (IE) D Acc EUR,689.900,18218.88,2026-04-30
2,IE00BYX5NX33,Fidelity MSCI World Index EUR P Acc,1332.135,17473.35,2026-04-30
3,LU0302296495,DNB Fund - Technology,7.136,11912.83,2026-04-30
4,ES0146309002,Horos Value Internacional FI,54.524,11708.41,2026-04-29
5,IE00BH6XSF26,Heptagon Kopernik Glb AllCp Eq AE € Acc,21.411,6682.66,2026-04-30
6,LU0273159177,Deutsche Invest I - Deutsche Invest I Gold and...,20.925,6365.70,2026-04-30
7,LU0329355670,Robeco Capital Growth Funds - Robeco QI Emergi...,15.899,5829.08,2026-04-30
8,LU1598719752,Cobas International P Acc EUR,32.883,5767.35,2026-04-29
9,LU2466448532,Echiquier Space,28.184,5482.07,2026-04-29



  Fondos en cartera:  18
  Capital Invertido:  125,028.28 €


---
## 3. Enriquecimiento de Datos Financieros

### Setup de proveedores + utilidades

Instanciamos cada proveedor de forma independiente. Cada bloque mide el tiempo por proveedor/ISIN para el benchmark final.

In [ ]:
import time
import importlib
from typing import Any, Dict, List

# Recargar módulos de servicios para evitar usar .pyc obsoletos
import app.services.finect_provider as _fp_mod
import app.services.ft_provider as _ft_mod
import app.services.data_providers as _dp_mod
importlib.reload(_fp_mod)
importlib.reload(_ft_mod)
importlib.reload(_dp_mod)

from app.services.data_providers import CompositeProvider, FMPProvider, MStarProvider, YFinanceProvider
from app.services.ft_provider import FTProvider
from app.services.finect_provider import FinectProvider

# ── Instanciar proveedores ──
finect = FinectProvider()
ft     = FTProvider()
mstar  = MStarProvider()
fmp    = FMPProvider()
yf     = YFinanceProvider()

print("✅ Proveedores inicializados (código recargado desde disco):")
print(f"   - FinectProvider  (INITIAL_STATE JSON: info, fees, sectores, regiones, alloc, holdings — 1 HTTP/fondo)")
print(f"   - FTProvider      (fallback: sectores, regiones, asset alloc, holdings)")
print(f"   - MStarProvider   (estrellas Morningstar, categoría, riesgo — única fuente de rating oficial)")
print(f"   - FMPProvider     (ETFs, expense_ratio — disponible: {fmp.available})")
print(f"   - YFinanceProvider (NAV, info básica, fallback)")

# ── Utilidades de formato homogéneo ──
def stars(rating) -> str:
    """Convierte rating numérico de Morningstar en ⭐."""
    try:
        n = int(float(rating))
        return "⭐" * n if 1 <= n <= 5 else str(rating)
    except (TypeError, ValueError):
        return "—"

def fmt_pct(val) -> str:
    """Formatea un valor como porcentaje o devuelve '—'."""
    if val is None:
        return "—"
    try:
        return f"{float(val):.2f}%"
    except (TypeError, ValueError):
        return str(val) if val else "—"

def nombre_corto(entry: Dict, max_len: int = 50) -> str:
    """Extrae nombre del fondo recortado."""
    n = entry.get("info", {}).get("name") or entry.get("morningstar", {}).get("name") or entry.get("isin", "?")
    return n[:max_len] + ("…" if len(str(n)) > max_len else "")

def timed_call(func, *args, **kwargs):
    """Ejecuta func midiendo tiempo. Devuelve (resultado, ms)."""
    t0 = time.perf_counter()
    try:
        result = func(*args, **kwargs)
        elapsed = (time.perf_counter() - t0) * 1000
        return result, round(elapsed, 1)
    except Exception:
        elapsed = (time.perf_counter() - t0) * 1000
        return None, round(elapsed, 1)

# ── Acumulador de benchmark ──
benchmark_rows: List[Dict[str, Any]] = []


✅ Proveedores inicializados:
   - FinectProvider  (INITIAL_STATE JSON: info, fees, sectores, regiones, alloc, holdings — 1 HTTP/fondo)
   - FTProvider      (fallback: sectores, regiones, asset alloc, holdings)
   - MStarProvider   (estrellas Morningstar, categoría, riesgo — única fuente de rating oficial)
   - FMPProvider     (ETFs, expense_ratio — disponible: True)
   - YFinanceProvider (NAV, info básica, fallback)


### Recolección de datos — todos los fondos × todos los proveedores

Un único loop sobre **todos los ISINs** que llama a cada proveedor midiendo tiempos.  
Los resultados se almacenan en `datos[isin]` para consultarse en cada bloque.

In [15]:
datos: Dict[str, Dict[str, Any]] = {}

n = len(isins_cartera)
for i, isin in enumerate(isins_cartera, 1):
    print(f"[{i:02d}/{n}] {isin} ... ", end="", flush=True)

    entry: Dict[str, Any] = {"isin": isin}

    # ── FINECT: una única petición HTTP por fondo → todos los bloques ──
    # La primera llamada descarga la página (~800ms).
    # Todas las siguientes (sectores, regiones, etc.) usan la caché en sesión → ~0ms.
    info_finect, t_finect = timed_call(finect.get_fund_info, isin)
    info_finect = info_finect or {}

    sectors_finect, _  = timed_call(finect.get_sector_weights, isin)
    regions_finect, _  = timed_call(finect.get_country_weights, isin)
    alloc_finect, _    = timed_call(finect.get_asset_allocation, isin)
    hold_finect, _     = timed_call(finect.get_holdings, isin)

    # ── 1. COMISIONES: Finect (principal) → FT → FMP ──
    info_ft, t_ft_info   = timed_call(ft.get_fund_info, isin)
    info_fmp, t_fmp_info = timed_call(fmp.get_fund_info, isin)

    info_ft  = info_ft  or {}
    info_fmp = info_fmp or {}

    # Fusionar: primer valor no vacío gana (Finect > FT > FMP)
    info_merged: Dict[str, Any] = {}
    for src in [info_finect, info_ft, info_fmp]:
        for k, v in src.items():
            if k not in info_merged or not info_merged[k] or info_merged[k] == "—":
                if v is not None and v != "" and v != isin:
                    info_merged[k] = v
    entry["info"] = info_merged

    benchmark_rows.append({"ISIN": isin, "Bloque": "Comisiones", "Proveedor": "Finect", "ms": t_finect,   "OK": bool(info_finect.get("name"))})
    benchmark_rows.append({"ISIN": isin, "Bloque": "Comisiones", "Proveedor": "FT",     "ms": t_ft_info,  "OK": bool(info_ft.get("name"))})
    benchmark_rows.append({"ISIN": isin, "Bloque": "Comisiones", "Proveedor": "FMP",    "ms": t_fmp_info, "OK": bool(info_fmp.get("name"))})

    # ── 2. SECTORES: Finect (principal) → FT → FMP ──
    sectors_ft, t_ft_sec   = timed_call(ft.get_sector_weights, isin)
    sectors_fmp, t_fmp_sec = timed_call(fmp.get_sector_weights, isin)
    entry["sectors"] = sectors_finect if sectors_finect else (sectors_ft if sectors_ft else (sectors_fmp or {}))

    benchmark_rows.append({"ISIN": isin, "Bloque": "Sectores", "Proveedor": "Finect", "ms": 0.0,        "OK": bool(sectors_finect)})
    benchmark_rows.append({"ISIN": isin, "Bloque": "Sectores", "Proveedor": "FT",     "ms": t_ft_sec,  "OK": bool(sectors_ft)})
    benchmark_rows.append({"ISIN": isin, "Bloque": "Sectores", "Proveedor": "FMP",    "ms": t_fmp_sec, "OK": bool(sectors_fmp)})

    # ── 3. REGIONES: Finect (principal) → FT → FMP ──
    regions_ft, t_ft_reg   = timed_call(ft.get_country_weights, isin)
    regions_fmp, t_fmp_reg = timed_call(fmp.get_country_weights, isin)
    entry["regions"] = regions_finect if regions_finect else (regions_ft if regions_ft else (regions_fmp or {}))

    benchmark_rows.append({"ISIN": isin, "Bloque": "Regiones", "Proveedor": "Finect", "ms": 0.0,         "OK": bool(regions_finect)})
    benchmark_rows.append({"ISIN": isin, "Bloque": "Regiones", "Proveedor": "FT",     "ms": t_ft_reg,   "OK": bool(regions_ft)})
    benchmark_rows.append({"ISIN": isin, "Bloque": "Regiones", "Proveedor": "FMP",    "ms": t_fmp_reg,  "OK": bool(regions_fmp)})

    # ── 4. ASSET ALLOCATION: Finect (principal) → FT → YFinance ──
    alloc_ft, t_ft_alloc = timed_call(ft.get_asset_allocation, isin)
    alloc_yf, t_yf_alloc = timed_call(yf.get_asset_allocation, isin)
    entry["asset_allocation"] = alloc_finect if alloc_finect else (alloc_ft if alloc_ft else (alloc_yf or {}))

    benchmark_rows.append({"ISIN": isin, "Bloque": "Asset Alloc", "Proveedor": "Finect",  "ms": 0.0,         "OK": bool(alloc_finect)})
    benchmark_rows.append({"ISIN": isin, "Bloque": "Asset Alloc", "Proveedor": "FT",      "ms": t_ft_alloc,  "OK": bool(alloc_ft)})
    benchmark_rows.append({"ISIN": isin, "Bloque": "Asset Alloc", "Proveedor": "YFinance","ms": t_yf_alloc,  "OK": bool(alloc_yf)})

    # ── 5. PORTFOLIO (Holdings): Finect (principal) → FT → YFinance → FMP ──
    hold_ft, t_ft_hold   = timed_call(ft.get_holdings, isin)
    hold_yf, t_yf_hold   = timed_call(yf.get_holdings, isin)
    hold_fmp, t_fmp_hold = timed_call(fmp.get_holdings, isin)

    best_hold = None
    for h in [hold_finect, hold_ft, hold_yf, hold_fmp]:
        if h is not None and not h.empty:
            if best_hold is None or len(h) > len(best_hold):
                best_hold = h
    entry["holdings"] = best_hold

    benchmark_rows.append({"ISIN": isin, "Bloque": "Portfolio", "Proveedor": "Finect",  "ms": 0.0,         "OK": hold_finect is not None and not hold_finect.empty})
    benchmark_rows.append({"ISIN": isin, "Bloque": "Portfolio", "Proveedor": "FT",      "ms": t_ft_hold,  "OK": hold_ft  is not None and not hold_ft.empty})
    benchmark_rows.append({"ISIN": isin, "Bloque": "Portfolio", "Proveedor": "YFinance","ms": t_yf_hold,  "OK": hold_yf  is not None and not hold_yf.empty})
    benchmark_rows.append({"ISIN": isin, "Bloque": "Portfolio", "Proveedor": "FMP",     "ms": t_fmp_hold, "OK": hold_fmp is not None and not hold_fmp.empty})

    # ── 6. ESTRELLAS MORNINGSTAR ──
    ms_info, t_ms = timed_call(mstar.get_fund_info, isin)
    ms_info = ms_info or {}
    entry["morningstar"] = ms_info

    # Fusionar nombre de MStar si falta
    if "name" not in info_merged and ms_info.get("name"):
        info_merged["name"] = ms_info["name"]

    benchmark_rows.append({"ISIN": isin, "Bloque": "Estrellas", "Proveedor": "Morningstar", "ms": t_ms, "OK": bool(ms_info.get("overallMorningstarRating"))})

    datos[isin] = entry

    # Resumen inline
    cov = []
    if info_merged.get("name"): cov.append("info")
    if ms_info.get("overallMorningstarRating"): cov.append("⭐")
    if entry["sectors"]: cov.append("sec")
    if entry["regions"]: cov.append("reg")
    if entry["asset_allocation"]: cov.append("alloc")
    if entry["holdings"] is not None and not entry["holdings"].empty: cov.append("hold")
    print(", ".join(cov) if cov else "sin datos")

print(f"\n✅ Datos recolectados para {len(datos)} fondos.")


[01/18] ES0140794001 ... Cache manager initialized with base path: c:\Users\jaguirrepeman\OneDrive - Deloitte (O365D)\Documents\DS\Finance\backend\app\data\cache

--- Procesando: ES0140794001 (detailed)---
✓ Datos obtenidos desde caché para ES0140794001
Obteniendo datos completos de Morningstar para ES0140794001...
Recopilando datos completos para ES0140794001...
❌ Error general obteniendo datos para ES0140794001: Error 403
            for the api https://global.morningstar.com/api/v1/fr/stores/data-points/fields. Message : Forbidden.
  [Info] Usando datos históricos de Yahoo para ES0140794001
  ⚠️ Datos con error, no se cachean para ES0140794001 (detailed)
Cache manager initialized with base path: c:\Users\jaguirrepeman\OneDrive - Deloitte (O365D)\Documents\DS\Finance\backend\app\data\cache

--- Procesando: ES0140794001 (detailed)---
✓ Datos obtenidos para ES0140794001: 1248 días
Obteniendo datos completos de Morningstar para ES0140794001...
Recopilando datos completos para ES01407940

---
### 3.1 Comisiones

Tabla resumen con **todos los fondos**: nombre, gasto corriente, comisión de suscripción y fuente.  
Proveedor principal: **Finect** (extrae `INITIAL_STATE` JSON — la **primera** llamada descarga la página y rellena *todos* los bloques 3.1–3.5 a la vez con una sola petición HTTP).  
Fallback: **FT** (ongoing charge), **FMP** (expense_ratio para ETFs).


In [7]:
rows = []
for isin, entry in datos.items():
    info = entry["info"]

    nombre = nombre_corto(entry, 55)
    categoria = str(info.get("category") or entry["morningstar"].get("categoryName") or "—")[:40]

    # Gasto corriente — intentar varios campos
    gasto = (
        info.get("ongoing_charge")
        or info.get("comision_de_gestion")
        or entry["morningstar"].get("ongoingCostsOtherCosts")
        or info.get("expense_ratio")
    )
    comision_susc = info.get("initial_charge") or info.get("comision_de_suscripcion")

    rows.append({
        "ISIN": isin,
        "Nombre": nombre,
        "Categoría": categoria,
        "Gasto corriente": fmt_pct(gasto),
        "C. Suscripción": fmt_pct(comision_susc),
        "Fuente": info.get("source", "—"),
    })

df_comisiones = pd.DataFrame(rows).set_index("ISIN")
print(f"=== 3.1 COMISIONES ({len(df_comisiones)} fondos) ===\n")
display(df_comisiones)

=== 3.1 COMISIONES (18 fondos) ===



,Nombre,Categoría,Gasto corriente,C. Suscripción,Fuente
ISIN,,,,,
ES0140794001,Gamma Global A FI,Open Ended Investment Company,0.95%,--,Finect
ES0141116030,ES0141116030,—,—,—,FinancialTimes
ES0146309002,Horos Value Internacional FI,Open Ended Investment Company,1.90%,--,Finect
IE000ZYRH0Q7,iShares Developed World Index Fund (IE) S Acc EUR,Unit Trust,0.06%,--,Finect
IE00B88WFS66,Federated Hermes Asia ex-Japan Equity Fund Cla...,Open Ended Investment Company,1.59%,5.00%,Finect
IE00BD0NCM55,iShares Developed World Index Fund (IE) D Acc EUR,Unit Trust,0.12%,--,Finect
IE00BH6XSF26,Heptagon Fund ICAV - Kopernik Global All-Cap E...,Open Ended Investment Company,1.64%,3.00%,Finect
IE00BYX5MX67,Fidelity S&P 500 Index Fund EUR P Acc,Open Ended Investment Company,0.06%,--,Finect
IE00BYX5NX33,Fidelity MSCI World Index Fund EUR P Acc,Open Ended Investment Company,0.12%,--,Finect


---
### 3.2 Sectores

Distribución sectorial de cada fondo.  
Proveedor principal: **Finect** (`stock-sector` del `INITIAL_STATE`, caché en sesión → ~0ms adicionales).  
Fallback: **FT** (holdings page, tabla "Sector") → **FMP** (ETF sector weightings).  
Los fondos sin datos de sectores (ej. renta fija pura) aparecen como `—`.


In [8]:
print("=== 3.2 SECTORES ===\n")

# Tabla pivot: filas=ISIN, columnas=sectores únicos
all_sectors = set()
for entry in datos.values():
    all_sectors.update(entry["sectors"].keys())

if all_sectors:
    rows_sec = []
    for isin, entry in datos.items():
        row = {"ISIN": isin, "Fondo": nombre_corto(entry, 40)}
        for s in sorted(all_sectors):
            val = entry["sectors"].get(s)
            row[s] = fmt_pct(val)
        rows_sec.append(row)
    df_sec = pd.DataFrame(rows_sec).set_index("ISIN")
    display(df_sec)
else:
    print("  ⚠️  No se obtuvieron datos de sectores para ningún fondo.")

# Detalle individual
print("\n--- Detalle por fondo ---")
for isin, entry in datos.items():
    sectores = entry["sectors"]
    if sectores:
        print(f"\n📌 {isin} — {nombre_corto(entry)}")
        df_s = pd.DataFrame(
            [{"Sector": k, "Peso (%)": fmt_pct(v)} for k, v in sectores.items()]
        )
        df_s = df_s.sort_values("Peso (%)", ascending=False).reset_index(drop=True)
        display(df_s)
    else:
        print(f"📌 {isin} — {nombre_corto(entry)}: sin datos de sectores")

=== 3.2 SECTORES ===



,Fondo,Basic Materials,Communication Services,Consumer Cyclical,Consumer Defensive,Energy,Financial Services,Healthcare,Industrials,Other,Real Estate,Technology,Utilities
ISIN,,,,,,,,,,,,,
ES0140794001,Gamma Global A FI,—,—,—,—,—,—,—,—,—,—,—,—
ES0141116030,ES0141116030,—,—,—,—,—,—,—,—,—,—,—,—
ES0146309002,Horos Value Internacional FI,7.72%,5.51%,23.84%,—,9.79%,17.65%,—,10.67%,2.58%,2.54%,4.76%,2.17%
IE000ZYRH0Q7,iShares Developed World Index Fund (IE) …,3.55%,8.74%,9.39%,5.74%,3.94%,15.95%,9.85%,11.95%,4.69%,—,25.82%,—
IE00B88WFS66,Federated Hermes Asia ex-Japan Equity Fu…,7.65%,9.30%,18.41%,12.09%,1.28%,13.32%,—,7.89%,0.26%,1.69%,27.48%,—
IE00BD0NCM55,iShares Developed World Index Fund (IE) …,3.55%,8.74%,9.39%,5.74%,3.94%,15.95%,9.85%,11.95%,4.69%,—,25.82%,—
IE00BH6XSF26,Heptagon Fund ICAV - Kopernik Global All…,—,—,—,—,—,—,—,—,—,—,—,—
IE00BYX5MX67,Fidelity S&P 500 Index Fund EUR P Acc,—,10.75%,10.15%,4.80%,3.35%,11.86%,8.57%,8.16%,3.70%,—,35.86%,2.26%
IE00BYX5NX33,Fidelity MSCI World Index Fund EUR P Acc,3.33%,8.83%,9.23%,5.09%,3.97%,15.58%,8.75%,11.11%,4.39%,—,28.15%,—



--- Detalle por fondo ---
📌 ES0140794001 — Gamma Global A FI: sin datos de sectores
📌 ES0141116030 — ES0141116030: sin datos de sectores

📌 ES0146309002 — Horos Value Internacional FI


,Sector,Peso (%)
0,Energy,9.79%
1,Basic Materials,7.72%
2,Communication Services,5.51%
3,Technology,4.76%
4,Consumer Cyclical,23.84%
5,Other,2.58%
6,Real Estate,2.54%
7,Utilities,2.17%
8,Financial Services,17.65%
9,Industrials,10.67%



📌 IE000ZYRH0Q7 — iShares Developed World Index Fund (IE) S Acc EUR


,Sector,Peso (%)
0,Healthcare,9.85%
1,Consumer Cyclical,9.39%
2,Communication Services,8.74%
3,Consumer Defensive,5.74%
4,Other,4.69%
5,Energy,3.94%
6,Basic Materials,3.55%
7,Technology,25.82%
8,Financial Services,15.95%
9,Industrials,11.95%



📌 IE00B88WFS66 — Federated Hermes Asia ex-Japan Equity Fund Class R…


,Sector,Peso (%)
0,Communication Services,9.30%
1,Industrials,7.89%
2,Basic Materials,7.65%
3,Technology,27.48%
4,Consumer Cyclical,18.41%
5,Financial Services,13.32%
6,Consumer Defensive,12.09%
7,Real Estate,1.69%
8,Energy,1.28%
9,Other,0.26%



📌 IE00BD0NCM55 — iShares Developed World Index Fund (IE) D Acc EUR


,Sector,Peso (%)
0,Healthcare,9.85%
1,Consumer Cyclical,9.39%
2,Communication Services,8.74%
3,Consumer Defensive,5.74%
4,Other,4.69%
5,Energy,3.94%
6,Basic Materials,3.55%
7,Technology,25.82%
8,Financial Services,15.95%
9,Industrials,11.95%


📌 IE00BH6XSF26 — Heptagon Fund ICAV - Kopernik Global All-Cap Equit…: sin datos de sectores

📌 IE00BYX5MX67 — Fidelity S&P 500 Index Fund EUR P Acc


,Sector,Peso (%)
0,Healthcare,8.57%
1,Industrials,8.16%
2,Consumer Defensive,4.80%
3,Technology,35.86%
4,Other,3.70%
5,Energy,3.35%
6,Utilities,2.26%
7,Financial Services,11.86%
8,Communication Services,10.75%
9,Consumer Cyclical,10.15%



📌 IE00BYX5NX33 — Fidelity MSCI World Index Fund EUR P Acc


,Sector,Peso (%)
0,Consumer Cyclical,9.23%
1,Communication Services,8.83%
2,Healthcare,8.75%
3,Consumer Defensive,5.09%
4,Other,4.39%
5,Energy,3.97%
6,Basic Materials,3.33%
7,Technology,28.15%
8,Financial Services,15.58%
9,Industrials,11.11%


📌 LU0273159177 — DWS Invest Gold and Precious Metals Equities LC: sin datos de sectores

📌 LU0302296495 — DNB Fund - Technology A EUR (Acc)


,Sector,Peso (%)
0,Consumer Cyclical,9.02%
1,Technology,61.70%
2,Financial Services,6.58%
3,Communication Services,18.64%
4,Healthcare,0.00%
5,Real Estate,0.00%
6,Consumer Defensive,0.00%
7,Energy,0.00%
8,Utilities,0.00%



📌 LU0329355670 — Robeco QI Emerging Markets Active Equities D €


,Sector,Peso (%)
0,Consumer Cyclical,9.36%
1,Communication Services,9.13%
2,Basic Materials,7.54%
3,Industrials,6.90%
4,Healthcare,4.94%
5,Technology,35.94%
6,Financial Services,21.67%
7,Energy,1.98%
8,Consumer Defensive,1.84%
9,Other,1.78%


📌 LU0840158819 — Storm Fund II - Storm Bond Fund RC EUR: sin datos de sectores

📌 LU1598719752 — Cobas LUX SICAV - Cobas International Fund-P Acc E…


,Sector,Peso (%)
0,Healthcare,6.73%
1,Energy,29.94%
2,Consumer Cyclical,20.57%
3,Technology,2.59%
4,Real Estate,2.16%
5,Industrials,18.38%
6,Basic Materials,11.76%
7,Consumer Defensive,1.73%
8,Communication Services,0.44%
9,Other,0.00%


📌 LU1694789451 — DNCA Invest Alpha Bonds A EUR: sin datos de sectores
📌 LU1988110927 — Buy & Hold Luxembourg B&H Bond Class 1: sin datos de sectores

📌 LU2466448532 — Echiquier Space B


,Sector,Peso (%)
0,Communication Services,8.46%
1,Industrials,52.44%
2,Consumer Cyclical,4.22%
3,Technology,30.97%
4,Utilities,2.93%
5,Financial Services,0.00%
6,Healthcare,0.00%
7,Real Estate,0.00%
8,Consumer Defensive,0.00%



📌 LU3038481936 — Hamco SICAV - Global Value R EUR Acc


,Sector,Peso (%)
0,Communication Services,6.92%
1,Real Estate,5.55%
2,Consumer Cyclical,27.62%
3,Energy,2.85%
4,Financial Services,2.78%
5,Technology,2.54%
6,Other,2.25%
7,Industrials,16.47%
8,Consumer Defensive,12.88%
9,Basic Materials,11.22%


---
### 3.3 Regiones

Distribución geográfica por región/país.  
Proveedor principal: **Finect** (`regional-exposure`, caché en sesión → ~0ms adicionales).  
Fallback: **FT** (holdings page, tabla "Region") → **FMP** (ETF country weightings).


In [9]:
print("=== 3.3 REGIONES ===\n")

for isin, entry in datos.items():
    regiones = entry["regions"]
    print(f"🌍 {isin} — {nombre_corto(entry, 55)}")
    if regiones:
        df_r = pd.DataFrame(
            [{"Región / País": k, "Peso (%)": fmt_pct(v)} for k, v in regiones.items()]
        )
        df_r = df_r.sort_values("Peso (%)", ascending=False).reset_index(drop=True)
        display(df_r)
    else:
        print("   ⚠️ sin datos de regiones\n")

=== 3.3 REGIONES ===

🌍 ES0140794001 — Gamma Global A FI


,Región / País,Peso (%)
0,Greater Europe,6.53%
1,Eurozone,4.79%
2,Europe - ex Euro,1.41%
3,Americas,0.99%
4,Latin America,0.66%
5,Canada,0.33%
6,United Kingdom,0.33%
7,Greater Asia,0.27%
8,Australasia,0.27%


🌍 ES0141116030 — ES0141116030
   ⚠️ sin datos de regiones

🌍 ES0146309002 — Horos Value Internacional FI


,Región / País,Peso (%)
0,United Kingdom,7.46%
1,Canada,7.03%
2,United States,6.36%
3,Greater Europe,52.25%
4,Latin America,5.94%
5,Eurozone,40.05%
6,Europe - ex Euro,4.74%
7,Americas,19.33%
8,Greater Asia,0.00%


🌍 IE000ZYRH0Q7 — iShares Developed World Index Fund (IE) S Acc EUR


,Región / País,Peso (%)
0,Eurozone,8.92%
1,Americas,73.16%
2,Greater Asia,7.92%
3,United States,69.63%
4,Japan,6.12%
5,Europe - ex Euro,4.31%
6,United Kingdom,3.80%
7,Canada,3.53%
8,Greater Europe,17.03%
9,Australasia,1.80%


🌍 IE00B88WFS66 — Federated Hermes Asia ex-Japan Equity Fund Class R EUR …


,Región / País,Peso (%)
0,Greater Asia,2.55%
1,Japan,2.55%
2,Greater Europe,2.53%
3,Europe - ex Euro,2.53%
4,Americas,1.00%
5,United States,1.00%
6,Canada,0.00%
7,Latin America,0.00%
8,United Kingdom,0.00%


🌍 IE00BD0NCM55 — iShares Developed World Index Fund (IE) D Acc EUR


,Región / País,Peso (%)
0,Eurozone,8.92%
1,Americas,73.16%
2,Greater Asia,7.92%
3,United States,69.63%
4,Japan,6.12%
5,Europe - ex Euro,4.31%
6,United Kingdom,3.80%
7,Canada,3.53%
8,Greater Europe,17.03%
9,Australasia,1.80%


🌍 IE00BH6XSF26 — Heptagon Fund ICAV - Kopernik Global All-Cap Equity Fun…


,Región / País,Peso (%)
0,United States,9.90%
1,Canada,9.73%
2,Greater Europe,6.87%
3,Eurozone,6.87%
4,Latin America,5.94%
5,Greater Asia,3.55%
6,Japan,3.55%
7,Americas,25.57%


🌍 IE00BYX5MX67 — Fidelity S&P 500 Index Fund EUR P Acc


,Región / País,Peso (%)
0,Americas,98.93%
1,United States,98.93%
2,Greater Europe,0.42%
3,Europe - ex Euro,0.26%
4,Eurozone,0.13%
5,United Kingdom,0.03%
6,Canada,0.00%
7,Latin America,0.00%
8,Greater Asia,0.00%


🌍 IE00BYX5NX33 — Fidelity MSCI World Index Fund EUR P Acc


,Región / País,Peso (%)
0,Eurozone,8.35%
1,Americas,73.80%
2,United States,70.41%
3,Greater Asia,7.28%
4,Japan,5.57%
5,Europe - ex Euro,4.03%
6,United Kingdom,3.51%
7,Canada,3.39%
8,Greater Europe,15.89%
9,Australasia,1.71%


🌍 LU0273159177 — DWS Invest Gold and Precious Metals Equities LC


,Región / País,Peso (%)
0,Greater Asia,9.34%
1,Australasia,8.89%
2,Americas,75.22%
3,Canada,52.40%
4,Greater Europe,4.82%
5,United Kingdom,4.82%
6,United States,22.56%
7,Japan,0.45%
8,Latin America,0.26%


🌍 LU0302296495 — DNB Fund - Technology A EUR (Acc)


,Región / País,Peso (%)
0,Europe - ex Euro,9.24%
1,Americas,64.44%
2,United States,64.44%
3,Greater Europe,23.29%
4,Greater Asia,2.07%
5,Japan,2.07%
6,Eurozone,12.33%
7,Middle East,1.72%


🌍 LU0329355670 — Robeco QI Emerging Markets Active Equities D €


,Región / País,Peso (%)
0,Americas,8.61%
1,Latin America,8.04%
2,Greater Europe,6.60%
3,Middle East,5.94%
4,Eurozone,0.66%
5,United States,0.57%
6,Greater Asia,0.00%


🌍 LU0840158819 — Storm Fund II - Storm Bond Fund RC EUR


,Región / País,Peso (%)
0,Americas,0.00%
1,United States,0.00%
2,Canada,0.00%
3,Latin America,0.00%
4,Greater Asia,0.00%
5,Greater Europe,0.00%
6,United Kingdom,0.00%
7,Eurozone,0.00%
8,Europe - ex Euro,0.00%
9,Middle East,0.00%


🌍 LU1598719752 — Cobas LUX SICAV - Cobas International Fund-P Acc EUR


,Región / País,Peso (%)
0,United States,8.70%
1,Greater Europe,57.10%
2,Latin America,4.47%
3,Eurozone,33.45%
4,Europe - ex Euro,13.59%
5,Americas,13.17%
6,United Kingdom,10.06%
7,Greater Asia,0.00%


🌍 LU1694789451 — DNCA Invest Alpha Bonds A EUR


,Región / País,Peso (%)
0,Greater Europe,0.08%
1,Eurozone,0.05%
2,United Kingdom,0.02%
3,Europe - ex Euro,0.01%
4,Americas,0.00%
5,United States,0.00%
6,Canada,0.00%
7,Greater Asia,0.00%


🌍 LU1988110927 — Buy & Hold Luxembourg B&H Bond Class 1


,Región / País,Peso (%)
0,Americas,0.00%
1,United States,0.00%
2,Canada,0.00%
3,Latin America,0.00%
4,Greater Asia,0.00%
5,Greater Europe,0.00%
6,United Kingdom,0.00%
7,Eurozone,0.00%
8,Europe - ex Euro,0.00%
9,Middle East,0.00%


🌍 LU2466448532 — Echiquier Space B


,Región / País,Peso (%)
0,Americas,70.34%
1,United States,68.47%
2,United Kingdom,4.46%
3,Greater Asia,3.40%
4,Japan,3.40%
5,Greater Europe,21.08%
6,Eurozone,16.62%
7,Canada,1.87%
8,Latin America,0.00%
9,Australasia,0.00%


🌍 LU3038481936 — Hamco SICAV - Global Value R EUR Acc


,Región / País,Peso (%)
0,Eurozone,8.12%
1,Canada,5.18%
2,Greater Asia,3.96%
3,Japan,3.96%
4,Americas,21.05%
5,United Kingdom,2.15%
6,United States,15.87%
7,Greater Europe,10.27%


---
### 3.4 Asset Allocation

Desglose por clase de activo (Renta Variable, Renta Fija, Liquidez, etc.).  
Proveedor principal: **Finect** (`asset-allocation`, caché en sesión → ~0ms adicionales).  
Fallback: **FT** (holdings page, tabla "Type") → **YFinance** (asset_classes).


In [10]:
print("=== 3.4 ASSET ALLOCATION ===\n")

all_asset_classes = set()
for entry in datos.values():
    all_asset_classes.update(entry["asset_allocation"].keys())

rows_alloc = []
for isin, entry in datos.items():
    alloc = entry["asset_allocation"]
    if alloc:
        row = {"ISIN": isin, "Fondo": nombre_corto(entry, 45)}
        for ac in sorted(all_asset_classes):
            row[ac] = fmt_pct(alloc.get(ac))
        rows_alloc.append(row)

if rows_alloc:
    df_alloc = pd.DataFrame(rows_alloc).set_index("ISIN")
    print(f"Fondos con Asset Allocation disponible: {len(rows_alloc)} / {len(datos)}")
    display(df_alloc)
else:
    print("  ⚠️  No se obtuvieron datos de asset allocation.")

sin_datos = [isin for isin, e in datos.items() if not e["asset_allocation"]]
if sin_datos:
    print(f"\nSin datos de asset allocation: {sin_datos}")

=== 3.4 ASSET ALLOCATION ===

Fondos con Asset Allocation disponible: 17 / 18


,Fondo,Cash,Non-UK bond,Non-UK stock,Other,UK bond,UK stock
ISIN,,,,,,,
ES0140794001,Gamma Global A FI,8.81%,79.35%,8.38%,0.00%,3.13%,0.33%
ES0146309002,Horos Value Internacional FI,11.12%,0.00%,79.76%,1.65%,0.00%,7.46%
IE000ZYRH0Q7,iShares Developed World Index Fund (IE) S Acc…,0.32%,0.00%,95.86%,0.05%,0.00%,3.76%
IE00B88WFS66,Federated Hermes Asia ex-Japan Equity Fund Cl…,0.63%,0.00%,99.37%,0.00%,0.00%,0.00%
IE00BD0NCM55,iShares Developed World Index Fund (IE) D Acc…,0.32%,0.00%,95.86%,0.05%,0.00%,3.76%
IE00BH6XSF26,Heptagon Fund ICAV - Kopernik Global All-Cap …,22.52%,0.00%,73.61%,1.33%,0.00%,2.54%
IE00BYX5MX67,Fidelity S&P 500 Index Fund EUR P Acc,0.55%,0.00%,99.42%,0.00%,0.00%,0.03%
IE00BYX5NX33,Fidelity MSCI World Index Fund EUR P Acc,1.52%,0.00%,94.93%,0.04%,0.00%,3.51%
LU0273159177,DWS Invest Gold and Precious Metals Equities …,0.47%,0.00%,94.53%,0.18%,0.00%,4.82%



Sin datos de asset allocation: ['ES0141116030']


---
### 3.5 Portfolio (Top Holdings)

Principales posiciones en cartera de cada fondo (top 10).  
Proveedor principal: **Finect** (`portfolio.holdings`, caché en sesión → ~0ms adicionales).  
Fallback: **FT** (holdings tearsheet) → **YFinance** → **FMP**.


In [11]:
print("=== 3.5 PORTFOLIO — TOP HOLDINGS ===\n")

for isin, entry in datos.items():
    holdings = entry["holdings"]
    print(f"🏦 {isin} — {nombre_corto(entry, 55)}")
    if holdings is not None and not holdings.empty:
        h = holdings.copy()
        if "weight" in h.columns:
            h["weight"] = h["weight"].apply(lambda x: fmt_pct(x))
            h = h.sort_values("weight", ascending=False)
        display(h.head(10).reset_index(drop=True))
    else:
        print("   ⚠️  Sin datos de holdings disponibles.\n")

=== 3.5 PORTFOLIO — TOP HOLDINGS ===

🏦 ES0140794001 — Gamma Global A FI


,name,ticker,weight,market_value
0,Kosmos Energy Gta Holdings 5.625%,,4.33%,0.0
1,Kinetics 9.875% 13/11/2029,,4.01%,0.0
2,Golar LNG Limited 3.75%,,3.89%,0.0
3,Navios Maritime Partners 7.75 07/11/2030,,3.86%,0.0
4,Navios South American Logistics Inc. 4.4375%,,3.74%,0.0
5,International Petroleum 7.5% 10/10/2030,,3.49%,0.0
6,Panoro Energy ASA 5.125%,,3.39%,0.0
7,Bluenord ASA 6%,,3.08%,0.0
8,Kosmos Energy Ltd 3.75%,,2.67%,0.0
9,Performance Shipping Inc 4.9375%,,2.61%,0.0


🏦 ES0141116030 — ES0141116030
   ⚠️  Sin datos de holdings disponibles.

🏦 ES0146309002 — Horos Value Internacional FI


,name,ticker,weight,market_value
0,Pluxee NVPLX:PAR,,5.94%,0.0
1,TGS ASATGS:OSL,,4.74%,0.0
2,Naspers LtdNPSNY:PKL,,4.00%,0.0
3,Ayvens SAAYV:PAR,,3.74%,0.0
4,Sun Hung Kai & Co Ltd86:HKG,,3.70%,0.0
5,Zegona Communications PLCZEG:LSE,,3.40%,0.0
6,PayPal Holdings IncPYPL:NSQ,,3.19%,0.0
7,Gestamp Automocion SAGEST:MCE,,3.06%,0.0
8,Acerinox SAACX:MCE,,2.95%,0.0
9,Onex CorpONEX:TOR,,2.72%,0.0


🏦 IE000ZYRH0Q7 — iShares Developed World Index Fund (IE) S Acc EUR


,name,ticker,weight,market_value
0,NVIDIA CorpNVDA:NSQ,,5.04%,0.0
1,Apple IncAAPL:NSQ,,4.54%,0.0
2,Microsoft CorpMSFT:NSQ,,3.25%,0.0
3,Amazon.com IncAMZN:NSQ,,2.36%,0.0
4,Alphabet IncGOOGL:NSQ,,2.12%,0.0
5,Alphabet Inc Class C,,1.77%,0.0
6,Broadcom IncAVGO:NSQ,,1.68%,0.0
7,Meta Platforms IncMETA:NSQ,,1.65%,0.0
8,Tesla IncTSLA:NSQ,,1.33%,0.0
9,Eli Lilly and CoLLY:NYQ,,0.99%,0.0


🏦 IE00B88WFS66 — Federated Hermes Asia ex-Japan Equity Fund Class R EUR …


,name,ticker,weight,market_value
0,Taiwan Semiconductor Manufacturing Co Ltd2330:TAI,,9.30%,0.0
1,Samsung Electronics Co Ltd,,7.27%,0.0
2,Tencent Holdings Ltd700:HKG,,6.98%,0.0
3,CP All PCLLVN:MUN,,3.61%,0.0
4,Bangkok Bank PCLBBL-R:SET,,3.13%,0.0
5,Samsung Fire & Marine Insurance Co Ltd,,3.00%,0.0
6,Samsung Electronics Co Ltd Participating Prefe...,,2.89%,0.0
7,Samsung Life Insurance Co Ltd,,2.73%,0.0
8,AAC Technologies Holdings Inc2018:HKG,,2.56%,0.0
9,Swatch Group AGUHR:SWX,,2.53%,0.0


🏦 IE00BD0NCM55 — iShares Developed World Index Fund (IE) D Acc EUR


,name,ticker,weight,market_value
0,NVIDIA CorpNVDA:NSQ,,5.04%,0.0
1,Apple IncAAPL:NSQ,,4.54%,0.0
2,Microsoft CorpMSFT:NSQ,,3.25%,0.0
3,Amazon.com IncAMZN:NSQ,,2.36%,0.0
4,Alphabet IncGOOGL:NSQ,,2.12%,0.0
5,Alphabet Inc Class C,,1.77%,0.0
6,Broadcom IncAVGO:NSQ,,1.68%,0.0
7,Meta Platforms IncMETA:NSQ,,1.65%,0.0
8,Tesla IncTSLA:NSQ,,1.33%,0.0
9,Eli Lilly and CoLLY:NYQ,,0.99%,0.0


🏦 IE00BH6XSF26 — Heptagon Fund ICAV - Kopernik Global All-Cap Equity Fun…


,name,ticker,weight,market_value
0,Valterra Platinum LtdVAL:JNB,,4.23%,0.0
1,Seabridge Gold IncSEA:TOR,,3.70%,0.0
2,LG Uplus Corp,,2.76%,0.0
3,K+S AGSDFX.N:GER,,2.47%,0.0
4,Range Resources CorpRRC:NYQ,,2.39%,0.0
5,Impala Platinum Holdings LtdIMP:JNB,,2.39%,0.0
6,LG Corp,,1.86%,0.0
7,KT CorpKT:NYQ,,1.72%,0.0
8,Paladin Energy LtdPDN:ASX,,1.65%,0.0
9,Golden Agri-Resources LtdE5H:SES,,1.55%,0.0


🏦 IE00BYX5MX67 — Fidelity S&P 500 Index Fund EUR P Acc


,name,ticker,weight,market_value
0,NVIDIA CorpNVDA:NSQ,,7.95%,0.0
1,Apple IncAAPL:NSQ,,6.48%,0.0
2,Microsoft CorpMSFT:NSQ,,5.20%,0.0
3,Amazon.com IncAMZN:NSQ,,4.03%,0.0
4,Broadcom IncAVGO:NSQ,,3.24%,0.0
5,Alphabet IncGOOGL:NSQ,,3.19%,0.0
6,Alphabet Inc Class C,,2.55%,0.0
7,Meta Platforms IncMETA:NSQ,,2.39%,0.0
8,Tesla IncTSLA:NSQ,,1.76%,0.0
9,Berkshire Hathaway IncBRK.B:NYQ,,1.38%,0.0


🏦 IE00BYX5NX33 — Fidelity MSCI World Index Fund EUR P Acc


,name,ticker,weight,market_value
0,NVIDIA CorpNVDA:NSQ,,5.56%,0.0
1,Apple IncAAPL:NSQ,,4.54%,0.0
2,Microsoft CorpMSFT:NSQ,,3.46%,0.0
3,Amazon.com IncAMZN:NSQ,,2.78%,0.0
4,Alphabet IncGOOGL:NSQ,,2.23%,0.0
5,Broadcom IncAVGO:NSQ,,2.15%,0.0
6,Alphabet Inc Class C,,1.86%,0.0
7,Meta Platforms IncMETA:NSQ,,1.66%,0.0
8,Tesla IncTSLA:NSQ,,1.24%,0.0
9,JPMorgan Chase & CoJPM:NYQ,,0.96%,0.0


🏦 LU0273159177 — DWS Invest Gold and Precious Metals Equities LC


,name,ticker,weight,market_value
0,Newmont CorporationNEM:NYQ,,9.26%,0.0
1,Agnico Eagle Mines LtdAEM:TOR,,8.71%,0.0
2,Anglogold Ashanti PLCAU:NYQ,,7.24%,0.0
3,Franco-Nevada CorpFNV:TOR,,6.79%,0.0
4,Gold Fields LtdGFI:JNB,,5.80%,0.0
5,Endeavour Mining plcEDV:LSE,,4.82%,0.0
6,Kinross Gold CorpK:TOR,,4.26%,0.0
7,Northern Star Resources LtdNST:ASX,,4.15%,0.0
8,Royal Gold IncRGLD:NSQ,,3.81%,0.0
9,B2Gold CorpBTO:TOR,,3.64%,0.0


🏦 LU0302296495 — DNB Fund - Technology A EUR (Acc)


,name,ticker,weight,market_value
0,Microsoft CorpMSFT:NSQ,,9.37%,0.0
1,Alphabet IncGOOGL:NSQ,,6.67%,0.0
2,NVIDIA CorpNVDA:NSQ,,6.28%,0.0
3,Amazon.com IncAMZN:NSQ,,6.20%,0.0
4,Meta Platforms IncMETA:NSQ,,5.49%,0.0
5,Visa IncV:NYQ,,4.21%,0.0
6,Atlassian CorpTEAM:NSQ,,3.71%,0.0
7,Telefonaktiebolaget LM EricssonERIC B:STO,,3.70%,0.0
8,SAP SESAPX:GER,,2.87%,0.0
9,Samsung Electronics Co LtdSMSN:LSE,,2.67%,0.0


🏦 LU0329355670 — Robeco QI Emerging Markets Active Equities D €


,name,ticker,weight,market_value
0,Samsung Electronics Co Ltd,,6.56%,0.0
1,Robeco QI Chinese A-share Active Equities Z €L...,,3.66%,0.0
2,SK Hynix Inc,,3.57%,0.0
3,Tencent Holdings Ltd700:HKG,,3.28%,0.0
4,MSCI Emerging Markets Index Future Mar 26,,3.17%,0.0
5,Alibaba Group Holding Ltd9988:HKG,,2.02%,0.0
6,Taiwan Semiconductor Manufacturing Co Ltd2330:TAI,,10.07%,0.0
7,Delta Electronics Inc2308:TAI,,1.27%,0.0
8,China Construction Bank Corp939:HKG,,1.11%,0.0
9,Vale SAVALE:NYQ,,1.03%,0.0


🏦 LU0840158819 — Storm Fund II - Storm Bond Fund RC EUR


,name,ticker,weight,market_value
0,Hofseth International AS,,2.53%,0.0
1,Nynas AB,,2.10%,0.0
2,EnQuest PLC,,2.01%,0.0
3,Shearwater Geoservices AS,,1.93%,0.0
4,International Petroleum Corp,,1.87%,0.0
5,Bluenord ASA,,1.76%,0.0
6,Golar LNG Limited,,1.68%,0.0
7,International Seaways Inc.,,1.39%,0.0
8,Tidewater Inc New,,1.36%,0.0
9,JS BidCo Oyj,,1.25%,0.0


🏦 LU1598719752 — Cobas LUX SICAV - Cobas International Fund-P Acc EUR


,name,ticker,weight,market_value
0,Golar LNG LtdGLNG:NSQ,,4.68%,0.0
1,CK Hutchison Holdings Ltd1:HKG,,4.25%,0.0
2,Atalaya Mining Copper SAE5S1:FRA,,4.06%,0.0
3,Danieli & C Officine Meccaniche SpADANR:MIL,,3.61%,0.0
4,BW Offshore LtdBWO:OSL,,3.41%,0.0
5,BW Energy LtdBWE:OSL,,3.22%,0.0
6,Wizz Air Holdings PLCWIZZ:LSE,,2.86%,0.0
7,Brava Energia SABRAV3:SAO,,2.84%,0.0
8,TGS ASATGS:OSL,,2.77%,0.0
9,Derichebourg SADBG:PAR,,2.51%,0.0


🏦 LU1694789451 — DNCA Invest Alpha Bonds A EUR


,name,ticker,weight,market_value
0,United States Treasury Notes 3.375%,,7.55%,0.0
1,Ostrum SRI Cash Plus SIFR001400R6M6:EUR,,4.72%,0.0
2,United Kingdom of Great Britain and Northern I...,,4.57%,0.0
3,United States Treasury Notes 1.875%,,4.50%,0.0
4,United States Treasury Notes 2.125%,,4.11%,0.0
5,European Union 2.75%,,3.40%,0.0
6,Spain (Kingdom of) 0.7%,,3.30%,0.0
7,United States Treasury Bonds 2.375%,,3.17%,0.0
8,10 Year Japanese Government Bond Future Mar 26,,-3.77%,0.0
9,10 Year Treasury Note Future June 26,,-12.30%,0.0


🏦 LU1988110927 — Buy & Hold Luxembourg B&H Bond Class 1


,name,ticker,weight,market_value
0,Erste Group Bank AG 7%,,5.25%,0.0
1,Atradius Credito Y Caucion Sa De Seguros Y Rea...,,4.85%,0.0
2,"Deutsche Bank Aktiengesellschaft, Frankfurt Br...",,4.68%,0.0
3,Eroski Sociedad Cooperativa 5.75%,,4.64%,0.0
4,CI Financial Corp 4.625%,,4.22%,0.0
5,International Petroleum Corp 7.5%,,3.54%,0.0
6,Intesa Sanpaolo S.p.A. 6.181%,,2.82%,0.0
7,Athora Holding Ltd. 5.875%,,2.71%,0.0
8,Tecnicas Reunidas SA 5.15%,,2.46%,0.0
9,Banque Internationale a Luxembourg S.A. 7.25%,,2.42%,0.0


🏦 LU2466448532 — Echiquier Space B


,name,ticker,weight,market_value
0,L3Harris Technologies IncLHX:NYQ,,6.12%,0.0
1,RTX CorpRTX:NYQ,,5.17%,0.0
2,Teledyne Technologies IncTDY:NYQ,,4.79%,0.0
3,Trimble IncTRMB:NSQ,,4.69%,0.0
4,BAE Systems PLCBA.:LSE,,4.46%,0.0
5,Leidos Holdings IncLDOS:NYQ,,4.40%,0.0
6,Kratos Defense and Security Solutions IncKTOS:NSQ,,4.33%,0.0
7,Amazon.com IncAMZN:NSQ,,4.22%,0.0
8,Planet Labs PBCPL:NYQ,,3.74%,0.0
9,Gilat Satellite Networks LtdGILT:TLV,,3.54%,0.0


🏦 LU3038481936 — Hamco SICAV - Global Value R EUR Acc


,name,ticker,weight,market_value
0,Sasol LtdSOL:JNB,,2.63%,0.0
1,Ayala CorpAC:PHS,,2.51%,0.0
2,GT Capital Holdings IncGTCAP:PHS,,2.46%,0.0
3,Ziff Davis IncZD:NSQ,,2.45%,0.0
4,Linamar CorpLNR:TOR,,2.20%,0.0
5,ManpowerGroup IncMAN:NYQ,,2.14%,0.0
6,Bermaz Auto BhdBAUTO:KLS,,2.13%,0.0
7,Indofood Sukses Makmur Tbk PTINDF:JKT,,2.08%,0.0
8,Sirius XM Holdings IncSIRI:NSQ,,1.98%,0.0
9,Orion SAOEC:NYQ,,1.80%,0.0


---
### 3.6 Estrellas Morningstar

Rating por estrellas, categoría y riesgo de cada fondo.  
Proveedor: **Morningstar** (mstarpy, única fuente de rating oficial).

In [12]:
rows_ms = []
for isin, entry in datos.items():
    ms = entry["morningstar"]
    rows_ms.append({
        "ISIN": isin,
        "Nombre": nombre_corto(entry, 50),
        "⭐ Rating": stars(ms.get("overallMorningstarRating")),
        "Rating 3Y": stars(ms.get("morningstarRatingFor3Year")),
        "Rating 5Y": stars(ms.get("morningstarRatingFor5Year")),
        "Categoría MStar": str(ms.get("categoryName") or "—")[:40],
        "Riesgo": str(ms.get("riskLevel") or ms.get("riskScore") or "—"),
    })

df_estrellas = pd.DataFrame(rows_ms).set_index("ISIN")
print(f"=== 3.6 ESTRELLAS MORNINGSTAR ({len(df_estrellas)} fondos) ===\n")
display(df_estrellas)

=== 3.6 ESTRELLAS MORNINGSTAR (18 fondos) ===



,Nombre,⭐ Rating,Rating 3Y,Rating 5Y,Categoría MStar,Riesgo
ISIN,,,,,,
ES0140794001,Gamma Global A FI,—,—,—,—,—
ES0141116030,ES0141116030,—,—,—,—,—
ES0146309002,Horos Value Internacional FI,—,—,—,—,—
IE000ZYRH0Q7,iShares Developed World Index Fund (IE) S Acc EUR,—,—,—,—,—
IE00B88WFS66,Federated Hermes Asia ex-Japan Equity Fund Cla...,—,—,—,—,—
IE00BD0NCM55,iShares Developed World Index Fund (IE) D Acc EUR,—,—,—,—,—
IE00BH6XSF26,Heptagon Fund ICAV - Kopernik Global All-Cap E...,—,—,—,—,—
IE00BYX5MX67,Fidelity S&P 500 Index Fund EUR P Acc,—,—,—,—,—
IE00BYX5NX33,Fidelity MSCI World Index Fund EUR P Acc,—,—,—,—,—


---
### 3.7 Benchmark: Rapidez × Completitud

Análisis comparativo de cada proveedor por bloque de datos.  
Combina **tiempo de respuesta** (ms) con **tasa de éxito** (% de fondos con datos).

Métricas:
- **Tiempo medio (ms)**: rapidez de respuesta por proveedor/bloque
- **Tasa de éxito (%)**: porcentaje de fondos para los que el proveedor devolvió datos
- **Score**: métrica combinada que pondera velocidad y completitud

> **Interpretación del tiempo de Finect**:
> Finect realiza **una sola petición HTTP** por fondo en el bloque "Comisiones" (get_fund_info) y cachea el modelo completo en sesión.
> Los bloques 3.2–3.5 (sectores, regiones, alloc, holdings) se muestran como **~0ms** porque leen directamente de la caché.
> El coste real de Finect es el tiempo medido en "Comisiones", amortizado entre todos los bloques que sirve.


In [16]:
df_bench = pd.DataFrame(benchmark_rows)

# ── Nota sobre Finect: "0ms" en bloques 3.2–3.5 porque usa la caché en sesión ──
# La única petición real de Finect se mide en "Comisiones" (get_fund_info).
# Para sectores/regiones/alloc/holdings el tiempo real de Finect = t_Comisiones/5 fondos
# El benchmark de 0ms para bloques cacheados se separa para no distorsionar el score.

print("=== 3.7 BENCHMARK: RAPIDEZ × COMPLETITUD ===\n")

# ── Tabla detallada: tiempo + éxito por proveedor y bloque ──
pivot = df_bench.groupby(["Bloque", "Proveedor"]).agg(
    Tiempo_medio_ms=("ms", "mean"),
    Tiempo_p90_ms=("ms", lambda x: x.quantile(0.9)),
    Éxitos=("OK", "sum"),
    Total=("OK", "count"),
).reset_index()
pivot["Tasa éxito (%)"] = (pivot["Éxitos"] / pivot["Total"] * 100).round(1)
pivot["Tiempo_medio_ms"] = pivot["Tiempo_medio_ms"].round(1)
pivot["Tiempo_p90_ms"] = pivot["Tiempo_p90_ms"].round(1)

# Score combinado: penaliza lentitud y premia completitud
pivot["Score"] = (pivot["Tasa éxito (%)"] * (1 / (1 + pivot["Tiempo_medio_ms"] / 1000))).round(1)

print("📊 Detalle por bloque y proveedor:")
display(pivot.sort_values(["Bloque", "Score"], ascending=[True, False]).set_index(["Bloque", "Proveedor"]))

# ── Resumen por proveedor (agregado) ──
print("\n📈 Resumen agregado por proveedor:")
resumen = df_bench.groupby("Proveedor").agg(
    Tiempo_medio_ms=("ms", "mean"),
    Llamadas_OK=("OK", "sum"),
    Llamadas_Total=("OK", "count"),
).reset_index()
resumen["Tasa éxito (%)"] = (resumen["Llamadas_OK"] / resumen["Llamadas_Total"] * 100).round(1)
resumen["Tiempo_medio_ms"] = resumen["Tiempo_medio_ms"].round(1)
resumen["Score"] = (resumen["Tasa éxito (%)"] * (1 / (1 + resumen["Tiempo_medio_ms"] / 1000))).round(1)
display(resumen.sort_values("Score", ascending=False).set_index("Proveedor"))

# ── Mejor proveedor por bloque ──
print("\n🏆 Mejor proveedor por bloque (mayor Score):")
best = pivot.loc[pivot.groupby("Bloque")["Score"].idxmax()][
    ["Bloque", "Proveedor", "Score", "Tiempo_medio_ms", "Tasa éxito (%)"]
]
display(best.set_index("Bloque"))

# ── Cobertura Finect por bloque ──
print("\n🔍 Cobertura de Finect por bloque:")
finect_bench = df_bench[df_bench["Proveedor"] == "Finect"].groupby("Bloque").agg(
    Éxitos=("OK", "sum"),
    Total=("OK", "count"),
    Tiempo_ms=("ms", "mean"),
).reset_index()
finect_bench["Tasa éxito (%)"] = (finect_bench["Éxitos"] / finect_bench["Total"] * 100).round(1)
finect_bench["Tiempo_ms"] = finect_bench["Tiempo_ms"].round(1)
finect_bench["Nota"] = finect_bench["Bloque"].apply(
    lambda b: "1 petición HTTP real" if b == "Comisiones" else "caché en sesión (~0ms)"
)
display(finect_bench.set_index("Bloque"))

t_finect_real = df_bench[(df_bench["Proveedor"] == "Finect") & (df_bench["Bloque"] == "Comisiones")]["ms"].mean()
print(f"\n⚡ Tiempo real medio de Finect (descarga + parse INITIAL_STATE): {t_finect_real:.0f} ms / fondo")
print(f"   → Amortizado entre 5 bloques: {t_finect_real/5:.0f} ms/bloque equivalente")


=== 3.7 BENCHMARK: RAPIDEZ × COMPLETITUD ===

📊 Detalle por bloque y proveedor:


Tiempo_medio_ms  Tiempo_p90_ms  Éxitos  Total  \
Bloque      Proveedor                                                    
Asset Alloc FT                    1381.4         1685.5      17     18   
            Finect                   0.0            0.0       0     18   
            YFinance                 0.0            0.0       0     18   
Comisiones  Finect                 864.3         1053.1      17     18   
            FT                    2542.4         3007.2      17     18   
            FMP                      0.0            0.0       0     18   
Estrellas   Morningstar          29823.8        31020.4       0     18   
Portfolio   FT                    1335.2         1580.3      17     18   
            FMP                      0.2            0.3       0     18   
            Finect                   0.0            0.0       0     18   
            YFinance                 0.3            0.4       0     18   
Regiones    FT                    1463.4         1810.4      17     18   
            FMP                      0.0            0.0       0     18   
            Finect                   0.0            0.0       0     18   
Sectores    FT                    1346.2         1618.4      11     18   
            FMP                      0.0            0.0       0     18   
            Finect                   0.0            0.0       0     18   

                         Tasa éxito (%)  Score  
Bloque      Proveedor                           
Asset Alloc FT                     94.4   39.6  
            Finect                  0.0    0.0  
            YFinance                0.0    0.0  
Comisiones  Finect                 94.4   50.6  
            FT                     94.4   26.6  
            FMP                     0.0    0.0  
Estrellas   Morningstar             0.0    0.0  
Portfolio   FT                     94.4   40.4  
            FMP                     0.0    0.0  
            Finect                  0.0    0.0  
            YFinance                0.0    0.0  
Regiones    FT                     94.4   38.3  
            FMP                     0.0    0.0  
            Finect                  0.0    0.0  
Sectores    FT                     61.1   26.0  
            FMP                     0.0    0.0  
            Finect                  0.0    0.0


📈 Resumen agregado por proveedor:


,Tiempo_medio_ms,Llamadas_OK,Llamadas_Total,Tasa éxito (%),Score
Proveedor,,,,,
FT,1613.7,79,90,87.8,33.6
Finect,172.9,17,90,18.9,16.1
FMP,0.0,0,72,0.0,0.0
Morningstar,29823.8,0,18,0.0,0.0
YFinance,0.2,0,36,0.0,0.0



🏆 Mejor proveedor por bloque (mayor Score):


,Proveedor,Score,Tiempo_medio_ms,Tasa éxito (%)
Bloque,,,,
Asset Alloc,FT,39.6,1381.4,94.4
Comisiones,Finect,50.6,864.3,94.4
Estrellas,Morningstar,0.0,29823.8,0.0
Portfolio,FT,40.4,1335.2,94.4
Regiones,FT,38.3,1463.4,94.4
Sectores,FT,26.0,1346.2,61.1



🔍 Cobertura de Finect por bloque:


,Éxitos,Total,Tiempo_ms,Tasa éxito (%),Nota
Bloque,,,,,
Asset Alloc,0,18,0.0,0.0,caché en sesión (~0ms)
Comisiones,17,18,864.3,94.4,1 petición HTTP real
Portfolio,0,18,0.0,0.0,caché en sesión (~0ms)
Regiones,0,18,0.0,0.0,caché en sesión (~0ms)
Sectores,0,18,0.0,0.0,caché en sesión (~0ms)



⚡ Tiempo real medio de Finect (descarga + parse INITIAL_STATE): 864 ms / fondo
   → Amortizado entre 5 bloques: 173 ms/bloque equivalente


---
### 3.8 Diagnóstico Finect — Cobertura real por ISIN

Comprueba directamente qué tipos de `breakdown` tiene cada ISIN en la caché de Finect.


In [18]:
print("=== 3.8 DIAGNÓSTICO FINECT POR ISIN ===\n")

import importlib
import app.services.finect_provider as fp_mod
# Forzar recarga del módulo para asegurar que tenemos la versión actualizada
importlib.reload(fp_mod)
from app.services.finect_provider import FinectProvider as FP2

finect2 = FP2()

rows_diag = []
for isin in isins_cartera:
    model = finect2._get_model(isin)
    if model is None:
        rows_diag.append({"ISIN": isin, "Nombre": "—", "breakdown_types": "⚠️ Sin URL en sitemap / fetch fallido",
                          "Sectores": 0, "Regiones": 0, "Alloc": 0, "Holdings": 0, "Fuente": "—"})
        continue

    breakdown = model.get("breakdown", [])
    btypes = [b.get("type", "?") for b in breakdown]
    
    sectors = fp_mod._extract_breakdown(model, "stock-sector")
    regions = fp_mod._extract_breakdown(model, "regional-exposure")
    alloc   = fp_mod._extract_breakdown(model, "asset-allocation")
    portfolio = model.get("portfolio", {})
    holdings_n = len(portfolio.get("holdings", []))
    
    rows_diag.append({
        "ISIN": isin,
        "Nombre": (model.get("name", "?") or "?")[:45],
        "breakdown_types": ", ".join(btypes) if btypes else "⚠️ breakdown vacío",
        "Sectores": len(sectors),
        "Regiones": len(regions),
        "Alloc":    len(alloc),
        "Holdings": holdings_n,
        "Fuente": model.get("category", {}).get("name", "—")[:25] if model.get("category") else "—",
    })

df_diag = pd.DataFrame(rows_diag).set_index("ISIN")
display(df_diag)

# Conclusión
fondos_con_breakdown = df_diag[df_diag["Sectores"] > 0]
print(f"\n✅ Fondos con sectores (Finect): {len(fondos_con_breakdown)} / {len(df_diag)}")
print(f"✅ Fondos con regiones (Finect): {len(df_diag[df_diag['Regiones'] > 0])} / {len(df_diag)}")
print(f"✅ Fondos con alloc (Finect):    {len(df_diag[df_diag['Alloc'] > 0])} / {len(df_diag)}")
print(f"✅ Fondos con holdings (Finect): {len(df_diag[df_diag['Holdings'] > 0])} / {len(df_diag)}")


=== 3.8 DIAGNÓSTICO FINECT POR ISIN ===



,Nombre,breakdown_types,Sectores,Regiones,Alloc,Holdings,Fuente
ISIN,,,,,,,
ES0140794001,Gamma Global FI,"asset-allocation, market-capitalization, regio...",7,10,3,10,Mixtos Defensivos EUR - G
ES0141116030,—,⚠️ Sin URL en sitemap / fetch fallido,0,0,0,0,—
ES0146309002,Horos Value Internacional FI,"asset-allocation, market-capitalization, regio...",11,12,3,10,RV Global Cap. Pequeña
IE000ZYRH0Q7,iShares Developed World Index Fund (IE),"asset-allocation, market-capitalization, regio...",11,14,4,10,RV Global Cap. Grande Ble
IE00B88WFS66,Federated Hermes Asia ex-Japan Equity Fund,"asset-allocation, market-capitalization, regio...",10,7,2,10,RV Asia (ex-Japón)
IE00BD0NCM55,iShares Developed World Index Fund (IE),"asset-allocation, market-capitalization, regio...",11,14,4,10,RV Global Cap. Grande Ble
IE00BH6XSF26,Heptagon Fund ICAV - Kopernik Global All-Cap,"asset-allocation, market-capitalization, regio...",11,15,5,10,RV Global Cap. Flexible
IE00BYX5MX67,Fidelity S&P 500 Index Fund,"asset-allocation, market-capitalization, regio...",11,7,2,10,RV USA Cap. Grande Blend
IE00BYX5NX33,Fidelity MSCI World Index Fund,"asset-allocation, market-capitalization, regio...",11,14,3,10,RV Global Cap. Grande Ble



✅ Fondos con sectores (Finect): 14 / 18
✅ Fondos con regiones (Finect): 15 / 18
✅ Fondos con alloc (Finect):    16 / 18
✅ Fondos con holdings (Finect): 16 / 18


---
### 3.9 Conclusión: ¿Usar Finect por defecto?

**Cobertura medida en esta cartera (18 fondos UCITS):**

| Bloque | Finect | FT | Ganador |
|---|---|---|---|
| Info / Comisiones / Stats | **94%** | 83% | ✅ Finect (+categoría, ratings, srri, alpha, beta, sharpe…) |
| Sectores | **78%** (14/18) | 61% (11/18) | ✅ Finect |
| Regiones | 83% (15/18) | **94%** (17/18) | FT por 1 fondo |
| Asset Allocation | 89% (16/18) | **94%** (17/18) | FT por 1 fondo |
| Holdings | 89% (16/18) | **94%** (17/18) | FT por 1 fondo |

**Velocidad real comparada:**

| Proveedor | Coste total para los 5 bloques | Observaciones |
|---|---|---|
| **Finect** | **~864 ms** (1 petición HTTP) | Una descarga cubre todos los bloques |
| **FT** | ~5.500 ms (4 peticiones × ~1.380ms) | 1 petición por bloque (holdings + summary) |

**Finect es ~6× más rápido que FT para cobertura equivalente.**

**Estrategia adoptada:**

> 🚀 **Finect como proveedor primario en todos los bloques.**  
> En 89% de los fondos de la cartera, una sola petición HTTP (~864ms) cubre info, fees, sectores, regiones, alloc, holdings y estadísticas.  
> Las llamadas subsiguientes al mismo ISIN son instantáneas (~0ms, caché en sesión).  
> Para el ~11% restante de fondos (sin datos Finect o no indexados): **FT → FMP/YFinance** como fallback.


---
## 4. Calculadora Fiscal — Retirada Optimizada

El `TaxOptimizer` analiza los lotes FIFO y sugiere qué vender para minimizar la ganancia patrimonial aflorada, según la legislación española.

In [ ]:
from app.services.tax_calculator import TaxOptimizer

cantidad_a_retirar = 50000
print(f"Simulando una retirada optimizada de {cantidad_a_retirar:,}€...")

optimizer = TaxOptimizer(portfolio)
plan = optimizer.optimize_withdrawal(cantidad_a_retirar)

print("\n=== Plan de Retirada ===")
print(f"  Cantidad Bruta Retirada:          {plan['withdrawn_amount']:>12,.2f} €")
print(f"  Ganancia Patrimonial Aflorada:    {plan['total_capital_gain']:>12,.2f} €")
print(f"  Impuestos Estimados a Pagar:      {plan['estimated_tax']:>12,.2f} €")
print(f"  Cantidad Neta en tu cuenta:       {plan['net_amount']:>12,.2f} €")

print("\n--- Ventas recomendadas (lotes específicos) ---")
rows_plan = []
for step in plan['plan']:
    rows_plan.append({
        "Fondo (ISIN)": step['Fondo'],
        "Fecha Compra": step['Fecha_Compra'].strftime('%Y-%m-%d'),
        "Participaciones": f"{step['Participaciones_Vendidas']:.4f}",
        "Importe Retirado": f"{step['Importe_Retirado']:,.2f} €",
    })
display(pd.DataFrame(rows_plan))

: 